bftengine/src/bftengine/ClientsManager.cpp

```diff
@@ -215,56 +223,45 @@ std::unique_ptr<ClientReplyMsg> ClientsManager::allocateNewReplyMsgAndWriteToSto
-  ConcordAssert(replyHeader->replyLength > 0);
-  ConcordAssert(replyHeader->replyLength + sizeof(ClientReplyMsgHeader) <= maxReplySize_);
-
-  uint32_t replyMsgSize = sizeof(ClientReplyMsgHeader) + replyHeader->replyLength;
-  uint32_t numOfPages = replyMsgSize / sizeOfReservedPage_;
-  uint32_t sizeLastPage = sizeOfReservedPage_;
-  if (replyMsgSize % sizeOfReservedPage_ != 0) {
-    numOfPages++;
-    sizeLastPage = replyMsgSize % sizeOfReservedPage_;
-  }
-  LOG_DEBUG(CL_MNGR, KVLOG(clientId, numOfPages, sizeLastPage));
-  auto r = std::make_unique<ClientReplyMsg>(myId_, replyHeader->replyLength);
-
-  // load reply message from reserved pages
-  for (uint32_t i = 0; i < numOfPages; i++) {
-    char* const ptrPage = r->body() + i * sizeOfReservedPage_;
-    const uint32_t sizePage = ((i < numOfPages - 1) ? sizeOfReservedPage_ : sizeLastPage);
-    stateTransfer_->loadReservedPage(resPageOffset() + firstPageId + i, sizePage, ptrPage);
-  }
+  for (int j = 0; j < maxNumOfReqsPerClient_; j++) {
+    stateTransfer_->loadReservedPage(
+        resPageOffset() + firstPageId + j * sizeOfReservedPage_, sizeOfReservedPage_, scratchPage_.data());
 
-  const auto& replySeqNum = r->reqSeqNum();
-  if (replySeqNum != requestSeqNum) {
-    if (maxNumOfReqsPerClient_ == 1) {
-      metric_reply_inconsistency_detected_.Get().Inc();
-      LOG_FATAL(CL_MNGR,
-                "The client reserved page does not contain a reply for specified request"
-                    << KVLOG(clientId, replySeqNum, requestSeqNum));
-      ConcordAssert(false);
+    ClientReplyMsgHeader* replyHeader = (ClientReplyMsgHeader*)scratchPage_.data();
+    ConcordAssert(replyHeader->msgType == MsgCode::ClientReply);
+    ConcordAssert(replyHeader->currentPrimaryId == 0);
+    ConcordAssert(replyHeader->replyLength > 0);
+    ConcordAssert(replyHeader->replyLength + sizeof(ClientReplyMsgHeader) <= maxReplySize_);
+    if (replyHeader->reqSeqNum != requestSeqNum) continue;
+
+    uint32_t replyMsgSize = sizeof(ClientReplyMsgHeader) + replyHeader->replyLength;
+    uint32_t numOfPages = replyMsgSize / sizeOfReservedPage_;
+    uint32_t sizeLastPage = sizeOfReservedPage_;
+    if (replyMsgSize % sizeOfReservedPage_ != 0) {
+      numOfPages++;
+      sizeLastPage = replyMsgSize % sizeOfReservedPage_;
+    }
+    LOG_DEBUG(CL_MNGR, KVLOG(clientId, numOfPages, sizeLastPage));
+    auto r = std::make_shared<ClientReplyMsg>(myId_, replyHeader->replyLength);
+    // load reply message from reserved pages
+    for (uint32_t i = 0; i < numOfPages; i++) {
+      char* const ptrPage = r->body() + i * sizeOfReservedPage_;
+      const uint32_t sizePage = ((i < numOfPages - 1) ? sizeOfReservedPage_ : sizeLastPage);
+      stateTransfer_->loadReservedPage(
+          resPageOffset() + firstPageId + (j * sizeOfReservedPage_) + i, sizePage, ptrPage);
     }
-    // YS TBD: Fix this for client batching with a proper ordering of incoming requests
-    LOG_INFO(CL_MNGR,
```

Comment:
In case we don't find a corresponding reply, log error.

Reply:
Added, thanks.